File: PHP/One/Transforms/InferTypesPlugins/ResolveNewCall.php
<?php

namespace One\Transforms\InferTypesPlugins\ResolveNewCall;

use One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use One\Ast\Expressions\Expression;
use One\Ast\Expressions\NewExpression;

class ResolveNewCalls extends InferTypesPlugin {
    function __construct() {
        parent::__construct("ResolveNewCalls");
        
    }
    
    function canTransform($expr) {
        return $expr instanceof NewExpression;
    }
    
    function transform($expr) {
        $newExpr = $expr;
        for ($i = 0; $i < count($newExpr->args); $i++) {
            $newExpr->args[$i]->setExpectedType($newExpr->cls->decl->constructor_->parameters[$i]->type);
            $newExpr->args[$i] = $this->main->runPluginsOn($newExpr->args[$i]);
        }
        $expr->setActualType($newExpr->cls);
        return $expr;
    }
}

File: CSharp/One/Transforms/InferTypesPlugins/ResolveNewCall.cs
using One.Ast;
using One.Transforms.InferTypesPlugins.Helpers;

namespace One.Transforms.InferTypesPlugins
{
    public class ResolveNewCalls : InferTypesPlugin {
        public ResolveNewCalls(): base("ResolveNewCalls")
        {
            
        }
        
        public override bool canTransform(Expression expr)
        {
            return expr is NewExpression;
        }
        
        public override Expression transform(Expression expr)
        {
            var newExpr = ((NewExpression)expr);
            for (int i = 0; i < newExpr.args.length(); i++) {
                newExpr.args.get(i).setExpectedType(newExpr.cls.decl.constructor_.parameters.get(i).type);
                newExpr.args.set(i, this.main.runPluginsOn(newExpr.args.get(i)));
            }
            expr.setActualType(newExpr.cls);
            return expr;
        }
    }
}